crates/rpc-types-eth/src/transaction/mod.rs

```diff
@@ -31,382 +28,224 @@ pub use receipt::AnyTransactionReceipt;
-    /// Max BaseFeePerGas the user is willing to pay.
-    #[cfg_attr(
-        feature = "serde",
-        serde(
-            default,
-            skip_serializing_if = "Option::is_none",
-            with = "alloy_serde::quantity::opt"
-        )
-    )]
-    pub max_fee_per_gas: Option<u128>,
-    /// The miner's tip.
-    #[cfg_attr(
-        feature = "serde",
-        serde(
-            default,
-            skip_serializing_if = "Option::is_none",
-            with = "alloy_serde::quantity::opt"
-        )
-    )]
-    pub max_priority_fee_per_gas: Option<u128>,
-    /// Configured max fee per blob gas for eip-4844 transactions
-    #[cfg_attr(
-        feature = "serde",
-        serde(
-            default,
-            skip_serializing_if = "Option::is_none",
-            with = "alloy_serde::quantity::opt"
-        )
-    )]
-    pub max_fee_per_blob_gas: Option<u128>,
-    /// Data
-    pub input: Bytes,
-    /// All _flattened_ fields of the transaction signature.
-    ///
-    /// Note: this is an option so special transaction types without a signature (e.g. <https://github.com/ethereum-optimism/optimism/blob/0bf643c4147b43cd6f25a759d331ef3a2a61a2a3/specs/deposits.md#the-deposited-transaction-type>) can be supported.
-    #[cfg_attr(feature = "serde", serde(flatten, skip_serializing_if = "Option::is_none"))]
-    pub signature: Option<Signature>,
-    /// The chain id of the transaction, if any.
-    #[cfg_attr(
-        feature = "serde",
-        serde(
-            default,
-            skip_serializing_if = "Option::is_none",
-            with = "alloy_serde::quantity::opt"
-        )
-    )]
-    pub chain_id: Option<ChainId>,
-    /// Contains the blob hashes for eip-4844 transactions.
-    #[cfg_attr(feature = "serde", serde(skip_serializing_if = "Option::is_none"))]
-    pub blob_versioned_hashes: Option<Vec<B256>>,
-    /// EIP2930
-    ///
-    /// Pre-pay to warm storage access.
-    #[cfg_attr(feature = "serde", serde(skip_serializing_if = "Option::is_none"))]
-    pub access_list: Option<AccessList>,
-    /// EIP2718
-    ///
-    /// Transaction type,
-    /// Some(4) for EIP-7702 transaction, Some(3) for EIP-4844 transaction, Some(2) for EIP-1559
-    /// transaction, Some(1) for AccessList transaction, None or Some(0) for Legacy
-    #[cfg_attr(
-        feature = "serde",
-        serde(
-            default,
-            rename = "type",
-            skip_serializing_if = "Option::is_none",
-            with = "alloy_serde::quantity::opt"
-        )
-    )]
-    #[doc(alias = "tx_type")]
-    pub transaction_type: Option<u8>,
-    /// The signed authorization list is a list of tuples that store the address to code which the
-    /// signer desires to execute in the context of their EOA and their signature.
-    #[cfg_attr(feature = "serde", serde(default, skip_serializing_if = "Option::is_none"))]
-    pub authorization_list: Option<Vec<SignedAuthorization>>,
+    // /// Transaction hash
+    // pub hash: B256,
```

Comment:
This is the last outstanding issue. For some reason, uncommenting this makes inner deserialization fail. Probably a subtle and annoying serde interaction 😮‍💨

Reply:
ah, it's because 
- `Signed` has an unskipped `hash` field
- the outer `Transaction` consumes the `"hash"` in deserialization
- there is no "hash" field left for `Signed` to deserialize

Reply:
so the problem becomes
- let `Signed` deserialize the hash
- make `hash` available on `Transaction` in a generic way

Reply:
OR 

- manually impl deserialize for Transaction to avoid consuming the hash

this seems easier and doesn't require introducing new bounds 👍

Reply:
should we just remove the `hash` from Transaction?

Reply:
that breaks the `TransactionResponse` trait. which would (e.g.) make it so that we can't index txns by hash without knowing the structure of the internal transaction. Which would suck

Reply:
recap:

the three options here:
1. remove hash from `Transaction` and `TransactionResponse`
2. Introduce a new trait bound in `Network` like this: `type TxEnvelope: NewTrait + Encodable2718`. This would expose the `tx_hash` and allow for `impl<T: NewTrait> TransactionResponse for Transaction<T>`
3. Modify the deserialize impl to not consume the `"hash"` key

#3 has no impact on our trait system and downstream users, and is non-breaking. so it seemed the best choice

Reply:
current problem is because `hash` is doubled when serializing 😮‍💨

Reply:
hmmmm serde and serde_json have no way to tell whether a map actually already contains a key while serializing, which means that we'd need to write our own serializer which we do _not_ want to do lol 

I'm now coming around to option 2 🤔

Reply:
> remove hash from Signed<T> is not practical because the method of calculating the hash relies on information about the tx encoding, which is underneath the abstraction. i.e. we don't have a trait that generically exposes the tx_hash calculation. the only way we calculate it is when constructing a Signed

I believe transaction hash is just `keccak256(tx_envelope.encode_2718())`? which should be possible to do for any envelope

Reply:
> I believe transaction hash is just keccak256(tx_envelope.encode_2718())? which should be possible to do for any envelope

will look into this today. at first blush, it looks like `trie_hash` is provided on the `Encodable2718` trait, so we should be able to remove from `Signed`. We still strongly want memoization tho so need to figure out how to keep that without interfering with serialization 🤔 

it also looks like 0c588bfeb31557680aa1bafaf06b7b60d7c9c9af lightly broke eip2718 abstraction by making type flags tx associated instead of envelope associated. We would now behave badly for networks where the type flag is different (e.g. `TxEip1559` is flag `17`, or where both `4` and `5` are `TxEip7702`, so we should probably open a separate issue for that

Reply:
> I believe transaction hash is just keccak256(tx_envelope.encode_2718())? which should be possible to do for any envelope

this is actually not the case for eip4844 with sidecar 😮‍💨  but it's straightforward to work around i think

Reply:
I'm wondering if we could solve this by adding a `TxEnvelope` trait which would have a `tx_hash` method? all normal variants already keep `hash` in `Signed` and we can just add a mandatory `hash` field to `Other` variant

I think https://github.com/alloy-rs/alloy/pull/1485 is reasonable as well because sealable and signable are 2 valid distinct properties of transaction (eg deposit transaction is only sealable). but would like to keep scope here smaller if possible as PR is already pretty big

Reply:
> I'm wondering if we could solve this by adding a TxEnvelope trait which would have a tx_hash method? all normal variants already keep hash in Signed and we can just add a mandatory hash field to Other variant

the `trie_hash` method already exists on `Encodable2718`, so we are covered there :)

Reply:
> the trie_hash method already exists on Encodable2718, so we are covered there :)

yeah but our impl of `Encodable2718` is not correct for `Other` variant?